### ISSUE TITLE
Potential Memory Leak: Memory Increases When Repeatedly Creating and Destroying Workers and Scheduler

### ISSUE THREAD

**Issue by ntoulasm** [naptha/tesseract.js#964]
**Tesseract.js version 5.1.1**

**Describe the bug**
 I believe there is a memory leak when repeatedly creating and destroying workers and schedulers in tesseract.js. Specifically, the memory consumption continues to increase over time, without being freed as expected.

**To Reproduce**
I have a piece of code that:

1. Creates a scheduler.
2. Creates several workers.
3. Adds a recognize job to the scheduler.
4. Destroys the scheduler and workers.
5. Repeats this process after 5 seconds.

```js
// scheduler.js
const { createScheduler, createWorker, OEM } = require('tesseract.js');

// --------------------------------------------------------------------------------------

const state = {
    scheduler: null,
    totalJobs: 0
};

// --------------------------------------------------------------------------------------

const MAX_WORKERS = Number(process.env.CHUNK_SIZE) || 8;
const JOB_LIMIT = 500 * MAX_WORKERS; 

// --------------------------------------------------------------------------------------

const create = async () => {
    const scheduler = createScheduler();
    const workers = await createWorkers(MAX_WORKERS);
    workers.forEach(scheduler.addWorker);
    return new Proxy(scheduler, {
        get(target, prop) {
            if (prop === 'addJob') {
                return async (...args) => {
                    ++state.totalJobs;
                    return await target[prop](...args);
                };
            }
            return target[prop];
        }
    });
};

/**
 * @async
 * @function
 * @returns {Promise<Tesseract.Scheduler>} - A promise that resolves to a Tesseract.js scheduler instance.
 */
module.exports.getScheduler = async () => {
    if (!state.scheduler) {
        state.scheduler = await create();
    }
    return state.scheduler;
};


module.exports.terminateSchedulerIfJobLimitExceeded = async () => {
    if (state.totalJobs < JOB_LIMIT) return;
    await terminate();
};

async function terminate() {
    await state.scheduler.terminate();
    state.scheduler = null;
    state.totalJobs = 0;
    global.gc?.();
}

// --------------------------------------------------------------------------------------

async function createWorkers(total) {
    return Promise.all(
        [...Array(total)].map(() =>
            createWorker('ell+eng', OEM.LSTM_ONLY, { langPath: './lib/ocr/tessdata' })
        )
    );
}

// --------------------------------------------------------------------------------------

const sharp = require('sharp');
const path = require('path');

async function main() {
    async function test() {
        console.log('.....');
        const scheduler = await module.exports.getScheduler(); 
        const image = await sharp(
            path.join(__dirname, '3fc9d821-c409-4c70-96e2-9c416c86ed20.jpg')
        ).toBuffer();
        await scheduler.addJob('recognize', image);
        await terminate();
        setTimeout(test, 5000);
    }
    test();
}

main();
```
The expected behavior is that memory would stabilize, but instead, the memory usage continues to increase, suggesting a memory leak. It seems that destroying the scheduler and workers is not releasing all the allocated resources.

Here is a screenshot, showing memory usage after 1-2 cycles:
![Screenshot 2024-09-30 145713](https://github.com/user-attachments/assets/8df0885f-80b4-437c-9d89-97b80066a997)
Here is another one after running for some time:
![Screenshot 2024-09-30 151606](https://github.com/user-attachments/assets/b9a2f30a-68a0-4ba0-823c-648168cb5505)

**Expected behavior**
The memory usage should increase during the creation and execution of the workers and the scheduler, but should be freed or stabilized once they are destroyed.

**Device Version:**
- Docker Version: 24.0.7, build afdd53b
- Operating System: Debian GNU/Linux 12 (bookworm)
- Node.js version: v18.20.4

**Additional context**
I understand that the usage pattern in this example is not typical or intended for tesseract.js. In my actual application, I reuse the same scheduler and workers, and reset them after processing 500 jobs, as suggested in the docs. However, I’ve observed that this also leads to increasing memory usage over time. The code provided here is a simplified example intended to demonstrate the memory growth issue more clearly.

**Comment by Balearica** [comment]
@ntoulasm I have a couple follow-ups regarding the example provided that should help us figure out whether a memory leak is the core issue.
1. Can you modify the example to no longer use recursive calls?
   1. I.e. define `test` as a top-level function and then call it `x` times within `main` using a loop.
   2. Recursive calls will inflate memory usage as they run, as every call adds a new frame to the call stack, and objects in earlier calls may not ever fall out of scope and be garbage collected.  
1. After editing, can you confirm that the increase over time continues (more or less) linearly, such that if run for long enough memory usage will reach >1GB?
   1. While the screenshots you attach do show an increase, as Node.js is a garbage collected language, I don't think seeing a ~200MB increase on the OS-level following a lot of memory-intensive operations is necessarily proof of a leak.

**Comment by ntoulasm** [comment]
Thank you for the follow-up.

First, I’d like to clarify that while the original example had a setTimeout call within the test function, I don’t believe it was true recursion since each call was independent and didn’t build up the call stack. Regardless, I’ve adjusted the code to remove the repeated calls and instead ran the function in a loop, as you suggested.

```js
async function main() {
    async function test() {
        const scheduler = await module.exports.getScheduler();
        const imagePath = imagePaths[Math.floor(Math.random() * imagePaths.length)];
        const image = await sharp(imagePath).toBuffer();
        await scheduler.addJob('recognize', image);
        await terminate();
    }

    while (true) {
        await test();
        await new Promise(resolve => setTimeout(resolve, 2000));
    }
}
```

After adjusting the example to use a loop instead of repeated calls, I’m still seeing similar memory growth. 

Memory usage before running the test:
![image](https://github.com/user-attachments/assets/49269197-9752-427f-99e4-d57c71c8cc29)

Memory usage after letting it run for a while:
![image](https://github.com/user-attachments/assets/8efce037-db62-45eb-9b44-ae0a06cf10d0)

While the increase is more modest in this simplified example(< 1GB), the behavior in my actual application is much more pronounced.

## Key Observations from the Actual Application:
**Number of workers**: I’m using 16 workers in total.

**Job limit**: The job limit is set to 16 workers * 500 jobs = 8000 jobs, after which I reset the workers and the scheduler.

**Memory usage**: The memory usage starts around 30 GB after creating the workers and running some jobs. However, after hitting the job limit, resetting the workers and scheduler, and running a job, it jumps to around 50 GB. This suggests that memory is not being released as expected when the workers and scheduler are terminated.

Memory usage during processing:
![image](https://github.com/user-attachments/assets/82e62cb1-cf37-4662-a774-c6a228e61286)

Memory usage after processing the first job after the first terminate:
![image](https://github.com/user-attachments/assets/f6e9b05a-bd06-43ea-8b8e-c12a112d6fed)

This behavior is significantly impacting the stability of the application, as memory usage grows far beyond what would be expected

**Comment by Balearica** [comment]
I tried running this code myself with one of our test images, and was unable to produce results consistent with a memory leak.  Memory usage rose over the first few minutes--which is expected when running any memory-intensive operation in a garbage-collected language--and then for the next half hour there was no change in memory.  Therefore, unless there is some reproducible example that can show memory growing linearly--e.g. a program where memory increases by 1GB every 5 minutes and it goes on forever--I am disinclined to believe there is a leak.

> The memory usage starts around 30 GB after creating the workers and running some jobs. However, after hitting the job limit, resetting the workers and scheduler, and running a job, it jumps to around 50 GB.

If your application is starting at ~30GB at startup then that's what you should focus on investigating.  Memory usage is normally nowhere close to that, even with 16 workers, as can be seen running a modified version of the [scheduler example code](https://github.com/naptha/tesseract.js/blob/master/examples/node/scheduler.js).  Off the top of my head, I would confirm that (1) all images you are passing in are of a reasonable size (a single giant 10,000x10,000 image will permanently increase the size of the worker due to Webassembly platform limitations) and (2) a large number of images are not being read into memory prior to being recognized.

**Comment by ntoulasm** [comment]
Thank you for the feedback. I'd like to provide more details to clarify my setup and see if it addresses the concern about memory usage.

Image Sizes: The dimensions of the images I process vary. The smallest are around 1200x1600 pixels, while the largest are around 4000x3000 pixels. Is this considered a reasonable size?

Image Loading: I process the images in chunks of 16 (1 image per worker) and load them just before the call to scheduler.recognize, so I am not loading a large number of images into memory prior to being recognized.

I will further investigate the issue to better understand why the workers memory usage is so high. Any additional suggestions you have for tracking down the source of the high memory usage would be greatly appreciated.

Lastly, when you mention that loading large images will permanently increase the size of the worker due to WebAssembly platform limitations, does this mean that even if the scheduler is terminated, the memory footprint from those large images will remain? Or is the memory expected to be reclaimed after resetting the scheduler and workers?

**Comment by ntoulasm** [comment]
Thanks to everyone for the feedback! After investigating further, it turns out the memory leak was caused by how I was managing the scheduler instances in my code. I was unintentionally creating multiple schedulers due to asynchronous calls that were racing to initialize the scheduler.

The root cause was that my getScheduler function would initiate a new scheduler whenever state.scheduler was null, leading to multiple schedulers being created in parallel. I have since refactored the code to ensure only one scheduler is created and reused across all promises.
